# Basic Gates Kata Workbook

**What is this workbook?** 
A workbook is a collection of problems, accompanied by solutions to them. The explanations focus on the logical steps required to solve a problem; they illustrate the concepts that need to be applied to come up with a solution to the problem, explaining the mathematical steps required.

Note that a workbook should not be the primary source of knowledge on the subject matter; it assumes that you've already read a tutorial or a textbook and that you are now seeking to improve your problem-solving skills. You should attempt solving the tasks of the respective kata first, and turn to the workbook only if stuck or for reinforcement. While a textbook emphasizes knowledge acquisition, a workbook emphasizes skill acquisition.

This workbook describes the solutions to the problems offered in the [Basic Gates Kata](./BasicGates.ipynb). 
Since the tasks are offered as programming problems, the explanations also cover some elements of Q# that might be non-obvious for a novitiate.

**What you should know for this workbook**

You should be familiar with the following concepts and associated techniques **prior to** beginning work on the Basic Gates Quantum Kata.

1. [Complex numbers](../tutorials/ComplexArithmetic/ComplexArithmetic.ipynb).
2. Basic linear algebra (multiplying column vectors by matrices), per the first part of [this tutorial](../tutorials/LinearAlgebra/LinearAlgebra.ipynb).
3. [The concept of qubit and its properties](../tutorials/Qubit/Qubit.ipynb).
4. [Single-qubit gates](../tutorials/SingleQubitGates/SingleQubitGates.ipynb).

You can also consult the [complete Quantum Katas learning path](https://github.com/microsoft/QuantumKatas#learning-path).

# Part 1. Single-Qubit Gates


### Task 1.1. State flip: $|0\rangle$ to $|1\rangle$ and vice versa

**Input:** A qubit in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

**Goal:**  Change the state of the qubit to $\alpha |1\rangle + \beta |0\rangle$.

**Example:**

If the qubit is in state $|0\rangle$, change its state to $|1\rangle$.

If the qubit is in state $|1\rangle$, change its state to $|0\rangle$.

## Solution

We can recognise that the Pauli X gate will change the state $|0\rangle$ to $|1\rangle$ and vice versa, and $\alpha |0\rangle + \beta |1\rangle$ to $\alpha |1\rangle + \beta |0\rangle$.

As a reminder, the Pauli X gate is defined by the following matrix:

$$
X = 
\begin{bmatrix}
   0 &  1\\
  1 &  0
\end{bmatrix}
$$

We can see how it affects, for example, the basis state $|0\rangle$:

$$X|0\rangle= 
\begin{bmatrix}
   0 &  1\\
  1 &  0
\end{bmatrix}
\begin{bmatrix}
   1\\
   0
\end{bmatrix}
=
\begin{bmatrix}
   0 \cdot 1 + 1 \cdot 0\\
   1 \cdot 1 + 0 \cdot 0
\end{bmatrix}
=
\begin{bmatrix}
   0\\
   1
\end{bmatrix}
=|1\rangle
$$

Similarly, we can consider the effect of the X gate on the superposition state $|\psi\rangle = 0.6|0\rangle + 0.8|1\rangle$:

$$X|\psi\rangle= 
\begin{bmatrix}
   0 &  1\\
  1 &  0
\end{bmatrix}
\begin{bmatrix}
   0.6\\
   0.8
\end{bmatrix}
=
\begin{bmatrix}
  0 \cdot 0.6  + 1 \cdot 0.8\\
   1 \cdot 0.6 + 0 \cdot 0.8
\end{bmatrix}
=
\begin{bmatrix}
   0.8\\
   0.6
\end{bmatrix}
=  0.8|0\rangle + 0.6|1\rangle
$$

In [1]:
%kata T101_StateFlip 

operation StateFlip (q : Qubit) : Unit is Adj+Ctl {
    X(q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.8 + 0𝑖
|1⟩	0.6 + 0𝑖

The actual state:


|0⟩	0.8 + 0𝑖
|1⟩	0.6 + 0𝑖

Success!

[Return to Task 1.1 of the Basic Gates kata.](./BasicGates.ipynb#Task-1.1.-State-flip:-$|0\rangle$-to-$|1\rangle$-and-vice-versa)

### Task 1.2. Basis change: $|0\rangle$ to $|+\rangle$ and $|1\rangle$ to $|-\rangle$ (and vice versa)

**Input**: A qubit in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

**Goal**:  Change the state of the qubit as follows:
* If the qubit is in state $|0\rangle$, change its state to $|+\rangle = \frac{1}{\sqrt{2}} \big(|0\rangle + |1\rangle\big)$.
* If the qubit is in state $|1\rangle$, change its state to $|-\rangle = \frac{1}{\sqrt{2}} \big(|0\rangle - |1\rangle\big)$.
* If the qubit is in superposition, change its state according to the effect on basis vectors.

## Solution

We can recognize that the Hadamard gate changes states $|0\rangle$ and $|1\rangle$ to $|+\rangle$ and $|-\rangle$, respectively, and vice versa.

As a reminder, the Hadamard gate is defined by the following matrix:

$$
\frac{1}{\sqrt{2}}\begin{bmatrix}
1 & 1 \\
1 & -1
\end{bmatrix}
$$

For example, we can work out $H|1\rangle$ as follows:

$$
H|1\rangle=
\frac{1}{\sqrt{2}}\begin{bmatrix}
   1 & 1 \\
   1 & -1
  \end{bmatrix}
 \begin{bmatrix}
   0\\
   1\\
  \end{bmatrix}
=
\frac{1}{\sqrt{2}}\begin{bmatrix}
   1 \cdot 0 + 1 \cdot 1 \\
   1 \cdot 0 + (-1) \cdot 1
  \end{bmatrix}
=
  \frac{1}{\sqrt{2}}\begin{bmatrix}
   1\\
   -1
  \end{bmatrix}
= \frac{1}{\sqrt{2}} \big(|0\rangle - |1\rangle\big) = |-\rangle
$$

Similarly, we can consider the effect of the Hadamard gate on the superposition state $|\psi\rangle = 0.6|0\rangle + 0.8|1\rangle$ (rounding the numbers to 4 decimal places):

$$
H|\psi⟩ = 
\frac{1}{\sqrt{2}}\begin{bmatrix}
   1 & 1 \\
   1 & -1
  \end{bmatrix}
 \begin{bmatrix}
   \alpha\\
   \beta\\
  \end{bmatrix}
 =
\frac{1}{\sqrt{2}}\begin{bmatrix}
   \alpha + \beta\\
   \alpha - \beta\\
  \end{bmatrix}
= 0.7071\begin{bmatrix}
   1.4\\
   -0.2\\
  \end{bmatrix}
= \begin{bmatrix}
   0.98994\\
   -0.14142\\
  \end{bmatrix}
= 0.9899|0\rangle - 0.1414|1\rangle
 $$

In [2]:
%kata T102_BasisChange 

operation BasisChange (q : Qubit) : Unit is Adj+Ctl {
    H(q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.9899494936611666 + 0𝑖
|1⟩	-0.1414213562373096 + 0𝑖

The actual state:


|0⟩	0.9899494936611665 + 0𝑖
|1⟩	-0.1414213562373096 + 0𝑖

Success!

[Return to Task 1.2 of the Basic Gates kata](./BasicGates.ipynb#Task-1.2.-Basis-change:-$|0\rangle$-to-$|+\rangle$-and-$|1\rangle$-to-$|-\rangle$-(and-vice-versa)).

## Task 1.3. Sign flip: $|+\rangle$  to $|-\rangle$  and vice versa.

**Input**: A qubit in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

**Goal** :  Change the qubit state to $\alpha |0\rangle - \beta |1\rangle$ (i.e. flip the sign of the $|1\rangle$ component of the superposition).


## Solution

The action of the Pauli Z gate is exactly what is required by this question.
This gate leaves the sign of the $|0\rangle$ component of the superposition unchanged but flips the sign of the $|1\rangle$ component of the superposition.

As a reminder, the Pauli Z gate is defined by the following matrix:

$$
Z = 
 \begin{bmatrix}
   1 & 0\\
   0 & -1
  \end{bmatrix}
 $$
 
Let's see its effect on the only computational basis state that it changes, $|1\rangle$:

$$
Z|1\rangle =
 \begin{bmatrix}
   1 & 0\\
   0 & -1
  \end{bmatrix}
 \begin{bmatrix}
   0\\
   1\\
  \end{bmatrix}
=
\begin{bmatrix}
   1 \cdot 0 + 0 \cdot1\\
   0 \cdot 1 +  -1 \cdot 1\\
  \end{bmatrix}
=
 \begin{bmatrix}
   0\\
   -1\\
  \end{bmatrix}
=
 -\begin{bmatrix}
   0\\
   1\\
  \end{bmatrix}
= -|1\rangle
$$

In general applying the Z gate to a single qubit superposition state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$ gives

$$
Z|\psi\rangle = 
 \begin{bmatrix}
   1 & 0 \\
   0 & -1
  \end{bmatrix}
 \begin{bmatrix}
   \alpha\\
   \beta\\
  \end{bmatrix}
  =
\begin{bmatrix}
   1\cdot\alpha + 0\cdot\beta\\
   0\cdot\alpha + -1\cdot\beta\\
  \end{bmatrix} 
  =
 \begin{bmatrix}
   \alpha\\
   -\beta\\
  \end{bmatrix}  
= \alpha |0\rangle -\beta |1\rangle
$$

In [3]:
%kata T103_SignFlip

operation SignFlip (q : Qubit) : Unit is Adj+Ctl {
    Z(q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.6 + 0𝑖
|1⟩	-0.8 + 0𝑖

The actual state:


|0⟩	0.6 + 0𝑖
|1⟩	-0.8 + 0𝑖

Success!

[Return to Task 1.3 of the Basic Gates kata](./BasicGates.ipynb#Task-1.3.-Sign-flip:-$|+\rangle$--to-$|-\rangle$--and-vice-versa.).

### Task 1.4. Amplitude change: $|0\rangle$ to $\cos{α} |0\rangle + \sin{α} |1\rangle$.

**Inputs:**

1. Angle α, in radians, represented as Double.
2. A qubit in state $|\psi\rangle = \beta |0\rangle + \gamma |1\rangle$.

**Goal:**  Change the state of the qubit as follows:
- If the qubit is in state $|0\rangle$, change its state to $\cos{α} |0\rangle + \sin{α} |1\rangle$.
- If the qubit is in state $|1\rangle$, change its state to $-\sin{α} |0\rangle + \cos{α} |1\rangle$.
- If the qubit is in superposition, change its state according to the effect on basis vectors.

## Solution

We can recognise that we need to use one of the rotation gates Rx, Ry, and Rz (named because they "rotate" the qubit state in the three dimensional space visualized as the Bloch sphere about the x, y, and z axes, respectively), since they involve angle parameters. Of these three gates, only Ry rotates the basis states $|0\rangle$ and $|1\rangle$ to have real amplitudes (the other two gates introduce complex coefficients).

As a reminder, 

$$
R_{y}(\theta) =
\begin{bmatrix}
   \cos \frac{\theta}{2} & -\sin \frac{\theta}{2}\\
   \sin \frac{\theta}{2} & \cos \frac{\theta}{2}
  \end{bmatrix}
$$

Let's see its effect on the $|0\rangle$ state:

$$
R_y(\theta)|0\rangle =
\begin{bmatrix}
   \cos \frac{\theta}{2} & -\sin \frac{\theta}{2}\\
   \sin \frac{\theta}{2} & \cos \frac{\theta}{2}
  \end{bmatrix}
\begin{bmatrix}
   1\\
   0\\
\end{bmatrix}
=
\begin{bmatrix}
   \cos \frac{\theta}{2}\cdot1 - \sin \frac{\theta}{2}\cdot0\\
   \sin \frac{\theta}{2}\cdot1 + \cos \frac{\theta}{2}\cdot0
  \end{bmatrix}
=
\begin{bmatrix}
   \cos \frac{\theta}{2}\\
   \sin \frac{\theta}{2}
  \end{bmatrix}
= \cos\frac{\theta}{2} |0\rangle + \sin\frac{\theta}{2} |1\rangle
$$

Recall that when applying a gate, you can tell what its matrix does to the basis states by looking at its columns: the first column of the matrix is the state into which it will transform the $|0\rangle$ state, and the second column is the state into which it will transfrom the $|1\rangle$ state.

In the example used by the testing harness we are given $\beta = 0.6, \gamma = 0.8$ and $\alpha = 1.0471975511965976 = \frac{\pi}{3}$.

Since $\cos \frac{\pi}{3} = 0.5$  and $\sin \frac{\pi}{3} = 0.8660$, working to 4 decimal places, we can compute:

$$
R_{y}(\theta) |\psi\rangle
=
 \begin{bmatrix}
   \cos \frac{\theta}{2} & -\sin \frac{\theta}{2}\\
   \sin \frac{\theta}{2} & \cos \frac{\theta}{2}
  \end{bmatrix} 
  \begin{bmatrix}
   \beta\\
  \gamma
 \end{bmatrix}
=
\begin{bmatrix}
   cos \frac{\theta}{2}\cdot\beta - sin \frac{\theta}{2}\cdot\gamma\\
   sin \frac{\theta}{2}\cdot\beta +cos \frac{\theta}{2}\cdot\gamma
  \end{bmatrix}
=  
 \begin{bmatrix}
   0.6\cdot\cos \frac{\pi}{3} -0.8\cdot\sin \frac{\pi}{3}\\
   0.6\cdot\sin \frac{\pi}{3} +0.8\cdot\cos \frac{\pi}{3}
  \end{bmatrix}
=
 \begin{bmatrix}
   0.3 - 0.6928\\
   0.5196 + 0.4
  \end{bmatrix}
=
  \begin{bmatrix}
   -0.3928\\
    0.9196
  \end{bmatrix}
$$ 

Notice that we used $\frac{\theta}{2} = \alpha$; this means that in the Q# code we need to pass the angle $\theta = 2\alpha$.

In [4]:
%kata T104_AmplitudeChange

operation AmplitudeChange (alpha : Double, q : Qubit) : Unit is Adj+Ctl {
    Ry(2.0 * alpha, q);
}

Applying amplitude change with alpha = 1.0471975511965976
The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	-0.3928203230275509 + 0𝑖
|1⟩	0.9196152422706633 + 0𝑖

The actual state:


|0⟩	-0.3928203230275509 + 0𝑖
|1⟩	0.9196152422706633 + 0𝑖

Success!

[Return to Task 1.4 of the Basic Gates kata](./BasicGates.ipynb#Task-1.4.-Amplitude-change:-$|0\rangle$-to-$\cos{α}-|0\rangle-+-\sin{α}-|1\rangle$.).

### Task 1.5. Phase flip

**Input:** A qubit in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

**Goal:** Change the qubit state to $\alpha |0\rangle + \color{red}i\beta |1\rangle$ (add a relative phase $i$ to $|1\rangle$ component of the superposition).


We can recognise that the S gate performs this particular relative phase addition of, in fact, $\frac{\pi}{4}$, to the $|1|\rangle$ component, and as a reminder

$$
S = 
 \begin{bmatrix}
   1 & 0\\
   0 & i
  \end{bmatrix}
 $$

Let's see the effect of this gate on the general superposition $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

$$
 \begin{bmatrix}
   1 & 0 \\
   0 & i
  \end{bmatrix}
 \begin{bmatrix}
   \alpha\\
   \beta\\
  \end{bmatrix}
  =
\begin{bmatrix}
   1\cdot\alpha + 0\cdot\beta\\
   0\cdot\alpha + i\cdot\beta
  \end{bmatrix} 
  =
 \begin{bmatrix}
   \alpha\\
   i\beta\\
  \end{bmatrix}  
$$

It is therefore easy to see that when It is therefore easy to see that when   $\psi= 0.6|0\rangle +  0.8|1\rangle, S(\psi) =  | 0.6|0\rangle + 0.8i|1\rangle$   

In [5]:
%kata T105_PhaseFlip

operation PhaseFlip (q : Qubit) : Unit is Adj+Ctl {
    S(q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.6 + 0𝑖
|1⟩	0 + 0.8𝑖

The actual state:


|0⟩	0 + 0.6𝑖
|1⟩	-0.8 + 0𝑖

Success!

## Solution 2

Please see Task 1.6, Phase Change, for an explanation of the following from the reference implementation:

In [6]:
       // alternatively R1(0.5 * PI(), q);

[Return to Task 1.5 of the Basic Gates kata](./BasicGates.ipynb#Task-1.5.-Phase-flip)

### Task 1.6. Phase change

**Inputs:**

1. Angle α, in radians, represented as Double.
2. A qubit in state $|\psi\rangle = \beta |0\rangle + \gamma |1\rangle$.

**Goal:**  Change the state of the qubit as follows:
- If the qubit is in state $|0\rangle$, don't change its state.
- If the qubit is in state $|1\rangle$, change its state to $e^{i\alpha} |1\rangle$.
- If the qubit is in superposition, change its state according to the effect on basis vectors: $\beta |0\rangle + \color{red}{e^{i\alpha}} \gamma |1\rangle$.

We know that:

$$
R1(\alpha)
= 
 \begin{bmatrix}
   1 & 0\\
   0 & \color{red}{e^{i\alpha}}
  \end{bmatrix}
 $$

So we have:

$$
 \begin{bmatrix}
   1 & 0 \\
   0 & \color{red}{e^{i\alpha}}
  \end{bmatrix}
 \begin{bmatrix}
   \alpha\\
   \beta\\
  \end{bmatrix}
  =
\begin{bmatrix}
   1.\alpha + 0.\beta\\
   0.\alpha + \color{red}{e^{i\alpha}}\beta
  \end{bmatrix} 
  =
 \begin{bmatrix}
   \alpha\\
   \color{red}{e^{i\alpha}}\beta
  \end{bmatrix}  
$$

Thus R1($\beta |0\rangle + \gamma |1\rangle$) = $\beta |0\rangle + \color{red}{e^{i\alpha}} \gamma |1\rangle$.

The result from the test harness is interesting  and there is something to notice here about qubit state amplitudes compared to measurements.
If you run the kata several times and  and examine the output, you'll notice that success is signalled even though the corresponding amplitudes of the desired and actual states are very different , in some cases there are even negative values.
So what's going on? The explanation is there in the ouput if you look carefully.
Rememeber that, according to Born's Rule,  we can only 'see' probabilities , not amplitudes.
When a qubit is measured, what is measured is the square of the modulus of the  amplitude 
(recall that the modulus squared of a complex number is that number multiplied by its complex conjugate)
Can we verify that assertion about amplitudes  below? Let's take a look.

In one run I got a desired amplitude of: $|0\rangle = 0.6000 + 0000i$, of $|1\rangle = -0.1389 +0,7878i$ and actual amplitudes of : $|0\rangle = 0.1042 + 0.5909$, and $|1\rangle = -0.7518 -0.2736i$.
If you do the arithmetic all the  $|0\rangle$ states come to .36 and all the $|1\rangle$ states come to 0.640062
For example $|-0.7518 -0.2736i|*|-0.7518 + 0.2736i| = .640062$, close enough to 64% when allowing for error in the amplitudes.
You could try other evaluations to verify the claim.
Success is therefore claimed on the basis that the desired and actual states are indistinguishable.

In [9]:
%kata T106_PhaseChange

operation PhaseChange (alpha : Double, q : Qubit) : Unit is Adj+Ctl {
   R1(alpha,q);
}

Applying phase change with alpha = 1.7453292519943295
The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.6 + 0𝑖
|1⟩	-0.13891854213354424 + 0.7878462024097665𝑖

The actual state:


|0⟩	-0.10418890660015818 + 0.5908846518073249𝑖
|1⟩	-0.7517540966287269 + -0.2736161146605349𝑖

Success!

Suppose now that $\alpha = \frac{\pi}{2} (=  0.5\pi)$
Then $e^{i\alpha}= \cos(\frac{\pi}{2}) + i\sin(\frac{pi}{2})$
And, since $\cos(\frac{\pi}{2})= 0$ and= $\sin(\frac{pi}{2}) = 1$, then we have that $\cos(\frac{\pi}{2}) + \sin(\frac{\pi}{2}) = i$, and  
$R1(\alpha)(\frac{\pi}{2}) = S$, as in the second solution to task 1.5, above.

[Return to Task 1.6 of the Basic Gates kata](./BasicGates.ipynb#Task-1.6.-Phase-Change)

### Task 1.7. Global phase change
**Input:** A qubit in state $|\psi\rangle = \beta |0\rangle + \gamma |1\rangle$.

**Goal**: Change the state of the qubit to $- \beta |0\rangle - \gamma |1\rangle$.

> Note: this change on its own is not observable - there is no experiment you can do on a standalone qubit to figure out whether it acquired the global phase or not. 
> However, you can use a controlled version of this operation to observe the global phase it introduces. 
> This is used in later katas as part of more complicated tasks.

<details>
  <summary><b>Need a hint? Click here</b></summary>
  Can you apply one of the rotation gates? Take a look at the functions in the Microsoft.Quantum.Math package to use a common mathematical constant and remember to import the package using the open directive.    
</details>

We recognise that a global phase can accomplished by using the R rotation gate with the exponentiated PauliI (identity) gate; at the same time we know that a global phase is not detectable and indeed that is has no physical meaning. It can be regarded as a mathematical aretefact. 
It disappers when you take a measurement of the state.  

To see this, consider a general state $\psi|\rangle$ with a global phase which can be written as e$^{i\alpha}$ which we can write as $|\phi\rangle = |\psi\rangle e^{i\alpha}$.  

Recall that using dirac notation we can write the measurement of $|\psi\rangle$ as $\langle\phi|\phi\rangle$ where we remember that the complex conjugate of $e^{i\alpha}$ is $e^{-i\alpha}$.  

This is  $e^ {-i\alpha}\langle\phi|e^{i\alpha}|\phi\rangle$ which is $e^{i\alpha}e^ {-i\alpha}\langle\phi|\phi\rangle$  which is $e^{-i\alpha+i\alpha }\langle\phi|\phi\rangle$ which is $e^{0}\langle\phi|\phi\rangle$ which is  $\langle\phi|\phi\rangle$ and our global phase has disappeared under measurement.

The R gate is defined as $R_{\mu}(\theta) = exp(\frac{\theta}{2}i\cdot\sigma_{\mu})$ wehere $\sigma_{\mu})$ is one of the Pauli gates I, X, Y or Z.  

For the problem at hand,  we have our general superposition state, $|\psi\rangle = \beta |0\rangle + \gamma |1\rangle$ , and our gate becomes R(PauliI,2$\pi$) = exp( i(2$\pi$/2)PauliI).  

Thus $R(|\psi\rangle)$ = PauliI$(cos\pi + isin\pi)$ and since $cos(\pi) = -1$   and   $sin(\pi) = 0$ we have that:   

$$
R(|\psi\rangle) 
=
 -1\begin{bmatrix}
   1 & 0 \\
   0 & 1
  \end{bmatrix}
 \begin{bmatrix}
   \beta\\
   \gamma\\
  \end{bmatrix}
=
 \begin{bmatrix}
   -1 & 0 \\
   0 & -1
  \end{bmatrix}
 \begin{bmatrix}
   \beta\\
   \gamma\\
  \end{bmatrix} 
=
 \begin{bmatrix}
   -1\cdot\beta + 0\cdot\gamma\\
    0\cdot\beta +  -1\cdot\gamma \\
  \end{bmatrix}
=
\begin{bmatrix}
   -\beta\\
   -\gamma\\
  \end{bmatrix}
=
- \beta |0\rangle - \gamma |1\rangle
$$

as required.
The solution below is for this result, but the test harness that runs when you execute the cell is for a *controlled global* phase change.

In [10]:
%kata T107_GlobalPhaseChange
open Microsoft.Quantum.Math;
operation GlobalPhaseChange (q : Qubit) : Unit is Adj+Ctl {
    R(PauliI, 2.0 * PI(), q);
    }

Showing effect of controlled-GlobalPhaseChange
The starting state:


|00⟩	0.4242640687119285 + 0𝑖
|01⟩	0.5656854249492381 + 0𝑖
|10⟩	0.4242640687119285 + 0𝑖
|11⟩	0.5656854249492381 + 0𝑖

The desired state:


|00⟩	0.4242640687119285 + 0𝑖
|01⟩	0.5656854249492381 + 0𝑖
|10⟩	-0.4242640687119285 + -5.19573633741296E-17𝑖
|11⟩	-0.5656854249492381 + -6.927648449883948E-17𝑖

The actual state:


|00⟩	0.4242640687119285 + 0𝑖
|01⟩	0.5656854249492381 + 0𝑖
|10⟩	-0.4242640687119285 + -5.19573633741296E-17𝑖
|11⟩	-0.5656854249492381 + -6.927648449883948E-17𝑖

Success!

[Return to Task 1.7 of the Basic Gates kata](./BasicGates.ipynb#Task-1.7.-Global-phase-change)